pkg/controller/certificates/signer/signer.go

```diff
@@ -89,37 +91,90 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
 }
 
 func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
+	// Ignore unapproved requests
 	if !certificates.IsCertificateRequestApproved(csr) {
 		return nil
 	}
-	csr, err := s.sign(csr)
+
+	// Special-case handling here in case we are talking to an older apiserver.
+	signerName := ""
+	if csr.Spec.SignerName != nil {
+		signerName = *csr.Spec.SignerName
+	} else {
+		signerName = capihelper.DefaultSignerNameFromSpec(&csr.Spec)
+	}
+
+	// Fast-path to avoid any additional processing if the CSRs signerName does
+	// not have a 'kubernetes.io/' prefix.
+	if !strings.HasPrefix(signerName, "kubernetes.io/") {
+		return nil
+	}
+
+	x509cr, err := capihelper.ParseCSR(csr.Spec.Request)
+	if err != nil {
+		return fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
+	}
+	if !requestValidForSignerName(x509cr, csr.Spec.Usages, signerName) {
+		// TODO: mark the CertificateRequest as being in a terminal state and
+		//  communicate to the user why the request has been refused.
+		return nil
+	}
+	cert, err := s.sign(x509cr, csr.Spec.Usages)
```

Comment:
I do not understand how this signing action makes sense with the actual signer name validation.  For example, in the case of `KubeAPIServerClientSignerName`, how do we make the API server honor the signed cert?  I would think we would need multiple `signer` controllers (one for each signer name) with (possibly) distinct private keys/CAs.

Reply:
The signing controller is configured to sign certificates using the apiserver's CA. Does the apiserver support specifying multiple roots of trust for verifying clients? If so, a user could configure a 'node CA' vs a 'user CA' if they wanted to revoke a certain subset of clients.

I think it does make sense to support specifying a CA to use per `kubernetes.io/*` signerName and @deads2k mentioned interest in this feature a little while ago too, but I'm uncertain whether it needs to be a pre-requisite for adding the `signerName` feature - if we add additional options to control this in future this can be handled via defaulting and it's not something outlined in the KEP currently.

Reply:
The `kubernetes.io/kubelet-serving` signer maps to `--kubelet-certificate-authority` (because this CA is a distinct flag, we can have a distinct private key for it).  We expect this signer to only be valid for serving certificates, but if we only use a single private key, then we are relying solely on cert usage checks.  We have no cryptographic isolation.

`kubernetes.io/kube-apiserver-client` and `kubernetes.io/kube-apiserver-client-kubelet` are a bit harder to separate since all we have is `--client-ca-file`.  @deads2k I might be misremembering but I seem to recall you doing something fancy with `--tls-sni-cert-key` and kubelets connecting over a particular network via IPs - but maybe that was for serving certs only.

In general I think we lack a way today to say "honor this CA for client certs but only if they match this particular shape (i.e. kubelet client cert vs. regular user client cert)."

`kubernetes.io/legacy-unknown` we can basically ignore.

Reply:
> For example, in the case of KubeAPIServerClientSignerName, how do we make the API server honor the signed cert?

That's the responsibility of the cluster administrator giving this controller the CA to use to sign and the apiserver the CA to use to verify.

> I would think we would need multiple signer controllers (one for each signer name) with (possibly) distinct private keys/CAs.

For this release, I think letting this signer continue to sign all the `kubernetes.io/*` names it understands is reasonable. If needed, we could allow subdivision of that similar to how we allow enabling/disabling specific controller loops. For example:
* `--cluster-signing-signer-names=*,-kubernetes.io/legacy-unknown`
* `--cluster-signing-signer-names=kubernetes.io/kube-apiserver-client-kubelet,kubernetes.io/kube-apiserver-client`

Reply:
> For this release, I think letting this signer continue to sign all the `kubernetes.io/*` names it understands with a single CA is reasonable.

Yeah I think it is probably the best we can do for now.